File: bunch.py
import os
import sys
import json
import argparse
import subprocess
import re
import shutil
from datetime import datetime
from pathlib import Path

def parse_args():
    parser = argparse.ArgumentParser(description='Batch download doodles for multiple topics')
    parser.add_argument('--type', type=str, default='topic',
                        help='Type of batch download (currently only "topic" is supported)')
    return parser.parse_args()

def load_topics():
    """Load topics from topics.json"""
    topics_file = Path(__file__).parent / 'topics.json'
    if not topics_file.exists():
        print(f"Error: topics.json not found at {topics_file}")
        sys.exit(1)
    
    with open(topics_file, 'r', encoding='utf-8') as f:
        topics = json.load(f)
    
    return topics

def is_already_downloaded(folder_path, topic_key):
    """
    Check if a topic has already been downloaded.
    Looks for a folder with pattern: {timestamp}_{topic_key} where timestamp is 14 consecutive digits (YYYYMMDDHHMMSS).
    """
    if not os.path.exists(folder_path):
        return False
    
    # Get the parent directory of the base folder
    parent_dir = Path(folder_path).parent
    
    # Build regex pattern: 14 digits followed by underscore and topic_key
    # Escape topic_key for regex
    escaped_key = re.escape(topic_key)
    pattern = re.compile(rf'^\d{{14}}_{escaped_key}$')
    
    try:
        for item in parent_dir.iterdir():
            if item.is_dir():
                # Check if name matches the pattern
                if pattern.match(item.name):
                    # Check if it's a valid doodle folder (contains images or images_info.json)
                    if (item / 'images_info.json').exists() or any(f.suffix in {'.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg'} for f in item.iterdir() if f.is_file()):
                        return True
    except Exception:
        pass
    
    return False

def cleanup_invalid_folders(images_folder, topics):
    """
    Delete directories that look like timestamped folders but are invalid.
    Scans: (1) parent of images_folder for timestamped folders, and (2) images_folder itself if it exists.
    Never deletes the 'images' folder or any valid doodle folders.
    """
    base_dir = Path(__file__).parent
    deleted = 0
    kept = 0
    
    # Precompile regex for valid doodle folders
    topic_keys = list(topics.keys())
    valid_patterns = {}
    for key in topic_keys:
        escaped_key = re.escape(key)
        valid_patterns[key] = re.compile(rf'^\d{{14}}_{escaped_key}$')
    
    def scan_and_clean(directory, is_nested=False):
        nonlocal deleted, kept
        if not directory.exists():
            return
        
        dir_name = directory.name
        
        # Check if directory name starts with 14 digits (timestamp format)
        if re.match(r'^\d{14}', dir_name):
            # This is a timestamped folder - check if valid
            is_valid = False
            for key, pattern in valid_patterns.items():
                if pattern.match(dir_name):
                    # Verify content
                    if (directory / 'images_info.json').exists() or any(f.suffix in {'.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg'} for f in directory.iterdir() if f.is_file()):
                        is_valid = True
                        break
            
            if is_valid:
                kept += 1
            else:
                # Delete the invalid folder
                try:
                    shutil.rmtree(directory)
                    print(f"  Deleted invalid folder: {directory}")
                    deleted += 1
                except Exception as e:
                    print(f"  Failed to delete {directory}: {e}")
            return  # Don't recurse into a timestamped folder (it's either kept or deleted)
        
        # If not a timestamped folder, and it's a directory, scan its contents
        # But skip 'images' folder at top level from being deleted (we still recurse into it)
        try:
            for item in directory.iterdir():
                if item.is_dir():
                    # At top level, skip the 'images' folder itself from deletion, but still recurse into it
                    if not is_nested and item.name == 'images':
                        scan_and_clean(item, is_nested=True)  # Recurse into images but don't delete it
                    else:
                        scan_and_clean(item, is_nested=False)
        except Exception:
            pass
    
    # Determine where to start scanning
    if images_folder.exists():
        print(f"Cleaning up: scanning {images_folder.parent} and {images_folder}")
        # Scan the parent directory (where renamed folders live)
        scan_and_clean(images_folder.parent, is_nested=False)
        # Also scan inside images folder itself (in case there are nested timestamped subfolders)
        scan_and_clean(images_folder, is_nested=False)
    else:
        # If images doesn't exist, scan the project directory
        parent_dir = base_dir
        print(f"Images folder not found, scanning project directory: {parent_dir}")
        scan_and_clean(parent_dir, is_nested=False)
    
    print(f"Cleanup: {deleted} invalid timestamped folders deleted, {kept} valid folders kept")
    print("="*60)

def main():
    args = parse_args()
    
    if args.type != 'topic':
        print(f"Error: Unsupported type '{args.type}'. Currently only 'topic' is supported.")
        sys.exit(1)
    
    # Load topics
    topics = load_topics()
    total_topics = len(topics)
    
    print(f"Found {total_topics} topics to process")
    print("="*60)
    
    # Get the directory where bunch.py is located
    base_dir = Path(__file__).parent
    doodles_script = base_dir / 'doodles.py'
    images_folder = base_dir / 'images'
    
    if not doodles_script.exists():
        print(f"Error: doodles.py not found at {doodles_script}")
        sys.exit(1)
    
    # Perform cleanup before starting batch
    cleanup_invalid_folders(images_folder, topics)
    
    # Track statistics
    skipped = 0
    completed = 0
    
    # Iterate through topics (English keys)
    for idx, topic_key in enumerate(topics.keys(), 1):
        topic_chinese = topics[topic_key]
        print(f"\n[{idx}/{total_topics}] Processing: {topic_key} ({topic_chinese})")
        
        # Construct query
        query = f"topic_tags={topic_key}"
        
        # Check if already downloaded
        if is_already_downloaded(images_folder, topic_key):
            print(f"  ⏭️  Skipping: Already downloaded (found existing folder with keyword '{topic_key}')")
            skipped += 1
            continue
        
        # Execute: python doodles.py --query={query}
        cmd = [sys.executable, str(doodles_script), '--query', query]
        
        print(f"  Running: {' '.join(cmd)}")
        try:
            result = subprocess.run(cmd, cwd=base_dir, check=False)
            
            if result.returncode == 0:
                print(f"  ✅ Completed: {topic_key}")
                completed += 1
            else:
                print(f"  ❌ Failed with exit code {result.returncode}")
                # Continue with next topic even if this one fails
        except Exception as e:
            print(f"  ❌ Error executing command: {e}")
    
    # Summary
    print("\n" + "="*60)
    print(f"Batch processing complete!")
    print(f"  Total topics: {total_topics}")
    print(f"  Skipped (already downloaded): {skipped}")
    print(f"  Completed this run: {completed}")
    print("="*60)

if __name__ == "__main__":
    main()

File: bunch.js
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// Load topics
const topicsPath = path.join(__dirname, 'topics.json');
const topics = JSON.parse(fs.readFileSync(topicsPath, 'utf-8'));
const topicKeys = Object.keys(topics);

// Precompile regex patterns for valid folders
const validPatterns = {};
for (const key of topicKeys) {
  // Match: {timestamp}_{sanitized_key}(_tmp)?
  const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, '\\$&').replace(/[/\\]|:|\*|\?|"|<|>|\|/g, '_');
  validPatterns[key] = new RegExp(`^\\d+_${escapedKey}(_tmp)?$`);
}

// Check if a folder represents a completed download for the given topic
function isAlreadyDownloaded(imagesFolder, topicKey) {
  if (!fs.existsSync(imagesFolder)) {
    return false;
  }

  const entries = fs.readdirSync(imagesFolder, { withFileTypes: true });

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const name = entry.name;
    const dirPath = path.join(imagesFolder, name);
    // Pattern: {timestamp}_{topic}
    if (validPatterns[topicKey].test(name) && !/\_tmp$/.test(dirPath)) {
      const hasImagesInfo = fs.existsSync(path.join(dirPath, 'images_info.json'));
      const hasImages = fs.readdirSync(dirPath).some(file =>
        ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg'].includes(path.extname(file).toLowerCase())
      );
      if (hasImagesInfo && hasImages) {
        return true;
      }
    }
  }

  return false;
}

// Cleanup invalid folders (folders that don't match any topic pattern)
function cleanupInvalidFolders(imagesFolder) {
  if (!fs.existsSync(imagesFolder)) {
    console.log('Images folder not found, skipping cleanup');
    return { deleted: 0, kept: 0 };
  }

  let deleted = 0;
  let kept = 0;
  const entries = fs.readdirSync(imagesFolder, { withFileTypes: true });

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const name = entry.name;

    // Skip if doesn't look like a timestamped folder
    if (!/^\d+/.test(name)) continue;

    // Check if it matches any valid topic pattern
    let isValid = false;
    for (const key of topicKeys) {
      if (validPatterns[key].test(name)) {
        const dirPath = path.join(imagesFolder, name);
        const hasImagesInfo = fs.existsSync(path.join(dirPath, 'images_info.json'));
        const hasImages = fs.readdirSync(dirPath).some(file =>
          ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.svg'].includes(path.extname(file).toLowerCase())
        );
        if (hasImagesInfo && hasImages) {
          isValid = true;
          break;
        }
      }
    }

    if (isValid) {
      kept++;
    } else {
      // Delete invalid folder
      try {
        const dirPath = path.join(imagesFolder, name);
        fs.rmSync(dirPath, { recursive: true });
        console.log(`  Deleted invalid folder: ${name}`);
        deleted++;
      } catch (e) {
        console.log(`  Failed to delete ${name}: ${e.message}`);
      }
    }
  }

  console.log(`Cleanup: ${deleted} invalid folders deleted, ${kept} valid folders kept`);
  console.log('='.repeat(60));
  return { deleted, kept };
}

// Cleanup _tmp folders for a specific topic (from previous failed runs)
function cleanupTopicTmpFolder(imagesFolder, topicKey) {
  if (!fs.existsSync(imagesFolder)) {
    return 0;
  }

  let deleted = 0;
  const entries = fs.readdirSync(imagesFolder, { withFileTypes: true });

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const name = entry.name;

    // Check if this is a _tmp folder for the given topic
    if (validPatterns[topicKey].test(name) && name.endsWith('_tmp')) {
      // Delete all _tmp folders for this topic (they represent incomplete/unfinalized downloads)
      const dirPath = path.join(imagesFolder, name);
      try {
        fs.rmSync(dirPath, { recursive: true });
        console.log(`  🗑️  Deleted previous _tmp folder: ${name}`);
        deleted++;
      } catch (e) {
        console.log(`  Failed to delete ${name}: ${e.message}`);
      }
    }
  }

  return deleted;
}

// Download a single topic
async function downloadTopic(topicKey, topicChinese, baseDir, imagesFolder) {
  return new Promise((resolve) => {
    const cmd = process.platform === 'win32' ? 'python' : 'python3';
    const doodlesScript = path.join(baseDir, 'doodles.py');
    const args = [doodlesScript, '--query', `topic_tags=${topicKey}`];

    const proc = spawn(cmd, args, {
      cwd: baseDir,
      stdio: ['ignore', 'inherit', 'inherit']
    });

    proc.on('close', (code) => {
      if (code === 0) {
        resolve({ topicKey, success: true });
      } else {
        resolve({ topicKey, success: false });
      }
    });
  });
}

// Main function
async function main() {
  const args = process.argv.slice(2);
  const options = {
    type: 'topic',
    workers: 3
  };

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--type' && args[i + 1]) {
      options.type = args[++i];
    } else if (args[i] === '--workers' && args[i + 1]) {
      options.workers = parseInt(args[++i], 10);
    }
  }

  if (options.type !== 'topic') {
    console.error(`Error: Unsupported type '${options.type}'. Currently only 'topic' is supported.`);
    process.exit(1);
  }

  console.log(`Found ${topicKeys.length} topics to process`);
  console.log(`Using ${options.workers} parallel workers`);
  console.log('Press Ctrl+C at any time to gracefully stop (once)');
  console.log('='.repeat(60));

  const baseDir = __dirname;
  const imagesFolder = path.join(baseDir, 'images');

  // Cleanup before starting
  cleanupInvalidFolders(imagesFolder);

  let skipped = 0;
  let completed = 0;
  let failed = 0;

  // Process topics with parallel workers
  const workerPromises = [];
  const activeWorkers = new Set();

  for (const topicKey of topicKeys) {
    // Wait if at max workers
    while (activeWorkers.size >= options.workers) {
      await new Promise(resolve => {
        for (const p of activeWorkers) {
          p.finally(() => {
            activeWorkers.delete(p);
            resolve();
          });
        }
      });
    }

    const topicChinese = topics[topicKey];
    console.log(`  ▶️  start downloading: ${topicKey} (${topicChinese})`);

    // Check if already downloaded
    if (isAlreadyDownloaded(imagesFolder, topicKey)) {
      console.log(`  ⏭️  Skipping ${topicKey}: Already downloaded`);
      skipped++;
      continue;
    }

    // Cleanup any previous failed _tmp folder for this topic before starting
    const cleanupCount = cleanupTopicTmpFolder(imagesFolder, topicKey);
    if (cleanupCount > 0) {
      console.log(`  🗑️  Cleaned up ${cleanupCount} previous failed _tmp folder(s)`);
    }

    const promise = downloadTopic(topicKey, topicChinese, baseDir, imagesFolder)
      .then(result => {
        if (result.success) {
          completed++;
        } else if (isAlreadyDownloaded(imagesFolder, topicKey)) {
          skipped++;
        } else {
          failed++;
        }
      })
      .catch(err => {
        console.log(`  ❌ Error for ${topicKey}: ${err.message || err}`);
        failed++;
      })
      .finally(() => {
        activeWorkers.delete(promise);
      });

    activeWorkers.add(promise);
    workerPromises.push(promise);
  }

  // Wait for all to complete
  await Promise.all(workerPromises);

  // Summary
  console.log('\n' + '='.repeat(60));
  console.log('Batch processing complete!');
  console.log(`  Total topics: ${topicKeys.length}`);
  console.log(`  Skipped (already downloaded): ${skipped}`);
  console.log(`  Completed this run: ${completed}`);
  console.log(`  Failed: ${failed}`);
  console.log('='.repeat(60));
}

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\n\n⚠️  Received interrupt signal, shutting down...');
  process.exit(0);
});
process.on('SIGTERM', () => {
  console.log('\n\n⚠️  Received terminate signal, shutting down...');
  process.exit(0);
});

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});